### PR TITLE
Set vf mac

### PR DESCRIPTION
**Pull request by ajitkhaparde**

### CODE REVIEW

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 1]
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
...
+	req.vnic_id_tbl_addr = rte_cpu_to_le_64(rte_mem_virt2phy(vnic_ids));
+
+	if (req.vnic_id_tbl_addr == 0) {
+		RTE_LOG(ERR, PMD,
+		"unable to map VNIC ID table address to physical memory\n");
+		rte_free(vnic_ids);
+		return -ENOMEM;
+	}
+	rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
+	HWRM_CHECK_RESULT;
```
Comment (stephen-hurd) [inline drivers/net/bnxt/bnxt_hwrm.c]: I don't think you can use HWRM_CHECK_RESULT if positive non-zero return values indicate success.

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 2]
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
...
+	HWRM_PREP(req, FUNC_VF_VNIC_IDS_QUERY, -1, resp_vf_vnic_ids);
+
+	req.vf_id = rte_cpu_to_le_16(bp->pf.first_vf_id + vf);
+	req.max_vnic_id_cnt = rte_cpu_to_le_32(bp->pf.total_vnics);
+	req.vnic_id_tbl_addr = rte_cpu_to_le_64(rte_mem_virt2phy(vnic_ids));
+
+	if (req.vnic_id_tbl_addr == 0) {
+		RTE_LOG(ERR, PMD,
+		"unable to map VNIC ID table address to physical memory\n");
+		rte_free(vnic_ids);
```
Comment (stephen-hurd) [inline drivers/net/bnxt/bnxt_hwrm.c]: This likely shouldn't be free()d in here.

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 3]
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
...
+	struct hwrm_func_vf_vnic_ids_query_input req = {0};
+	struct hwrm_func_vf_vnic_ids_query_output *resp =
+						bp->hwrm_cmd_resp_addr;
+	int rc;
+
+	/* First query all VNIC ids */
+	HWRM_PREP(req, FUNC_VF_VNIC_IDS_QUERY, -1, resp_vf_vnic_ids);
+
+	req.vf_id = rte_cpu_to_le_16(bp->pf.first_vf_id + vf);
+	req.max_vnic_id_cnt = rte_cpu_to_le_32(bp->pf.total_vnics);
```
Comment (stephen-hurd) [inline drivers/net/bnxt/bnxt_hwrm.c]: So we're assuming the size of the passed-in buffer?  Maybe we should have a separate parameter to define the size.

File: drivers/net/bnxt/rte_pmd_bnxt.c [hunk 4]
```diff
@@ -450,5 +451,81 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
...
+
+	/* Attach requested MAC address to the new l2_filter */
+	STAILQ_FOREACH(filter, &vf->filter, next) {
+		if (memcmp(addr, filter->l2_addr, ETHER_ADDR_LEN) == 0) {
+			RTE_LOG(ERR, PMD,
+				"Filter already exists for MAC addr\n");
+			return 0;
+		}
+	}
+	filter = bnxt_alloc_filter(bp);
```
Comment (stephen-hurd) [inline drivers/net/bnxt/rte_pmd_bnxt.c]: Where do these get free()d?

File: drivers/net/bnxt/rte_pmd_bnxt.c [hunk 5]
```diff
@@ -450,5 +451,81 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
...
+
+	for (sz = 0; sz < vnic_id_sz; sz += getpagesize())
+		rte_mem_lock_page(((char *)vnic_ids) + sz);
+
+	rc = bnxt_hwrm_func_vf_vnic_query(bp, vf_id, vnic_ids);
+	if (rc <= 0)
+		goto free;
+
+	memset(&vnic, 0, sizeof(struct bnxt_vnic_info));
+	vnic.fw_vnic_id = rte_le_to_cpu_16(vnic_ids[0]);
```
Comment (stephen-hurd) [inline drivers/net/bnxt/rte_pmd_bnxt.c]: You should use the default one, not the first one.

**Review by stephen-hurd — COMMENTED** [review]
I'm most worried about the cleanup question.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 2 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/bnxt_hwrm.c]: Let me do a single point of rte_free() in the calling function.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 1 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/bnxt_hwrm.c]: In this case, will bnxt_hwrm_send_message() return a positive non-zero value?
It is used like in any other place as earlier code.

File: drivers/net/bnxt/rte_pmd_bnxt.c (hunk 4 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/rte_pmd_bnxt.c]: There is no delete happening in the VFd, like an add.
I will add something in the course of normal cleanup?

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 1 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_hwrm.c]: bnxt_hwrm_send_message() returns -1 on failure, zero on success.

HWRM_CHECK_RESULT will cause the function it's in to return a positive error code from the response on failure (the error_code value from the response).

See bnxt_hwrm_func_qcfg_current_vf_vlan() for a function that returns a positive value on success.

File: drivers/net/bnxt/rte_pmd_bnxt.c (hunk 4 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/rte_pmd_bnxt.c]: VFd attempts to the filters when the VF is deleted by calling set_vf_rx_mac() with the on parameter set to false.  The actual implementation in sriov.c is clearly broken however and should likely be fixed.

But yes, adding cleanup to bnxt_free_all_filters() should be enough.

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 6]
```diff
@@ -2500,3 +2500,57 @@ int bnxt_hwrm_func_cfg_vf_set_vlan_anti_spoof(struct bnxt *bp, uint16_t vf)
...
+	} else if (resp->error_code) {
+		rc = rte_le_to_cpu_16(resp->error_code);
+		RTE_LOG(ERR, PMD, "hwrm_func_qcfg error %d\n", rc);
+		return -1;
+	}
+
+	return rte_le_to_cpu_32(resp->vnic_id_cnt);
+}
+
+int bnxt_hwrm_func_qcfg_vf_dflt_vnic_id(struct bnxt *bp, int vf)
```
Comment (stephen-hurd) [inline drivers/net/bnxt/bnxt_hwrm.c:2538]: This is the one that doesn't currently work, right?

Having this call bnxt_hwrm_func_vf_vnic_query() then loop through the vnics should work, then just this one place would need to be updated in the future.

File: drivers/net/bnxt/rte_pmd_bnxt.c [hunk 7]
```diff
@@ -452,3 +453,78 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
...
+		return rc;
+	}
+
+	for (sz = 0; sz < vnic_id_sz; sz += getpagesize())
+		rte_mem_lock_page(((char *)vnic_ids) + sz);
+
+	rc = bnxt_hwrm_func_vf_vnic_query(bp, vf_id, vnic_ids);
+	if (rc <= 0)
+		goto exit;
+	num_vnic_ids = rc;
```
Comment (stephen-hurd) [inline drivers/net/bnxt/rte_pmd_bnxt.c]: All of this could go in bnxt_hwrm_func_qcfg_vf_dflt_vnic_id() instead.

**Review by stephen-hurd — APPROVED** [review]
I suppose it's good enough for now.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 6 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/bnxt_hwrm.c:2538]: Done.

**Pull request merged**